File: code/src/ui/ui_element.cpp
#include "ui/ui_tree.h"
#include <algorithm>

using namespace tev::ui;

UIElement::UIElement(UITree* tree) : tree(tree), parent(nullptr) {}

UIElement::~UIElement() 
{
	this->on_destroyed();

	this->set_parent(nullptr);

	for (UIElement* child : this->children)
	{
		delete child;
	}

	this->children.clear();
}

void UIElement::on_added_to_tree() {}
void UIElement::on_removed_from_tree() {}

void UIElement::resize_children_capacity(size_t new_capacity)
{
	this->children.reserve(new_capacity);
}

void UIElement::add_child(UIElement* child) 
{
	if (child == this)
	{
		return;
	}

	if (child->get_parent() == this)
	{
		return;
	}

	auto it = std::find(
		this->children.begin(),
		this->children.end(),
		child
	);

	if (it == this->children.end())
	{
		this->children.emplace_back((*it));

		child->parent = this;

		child->on_added_to_tree();

		this->internal_add_child(child);
	}
}

void UIElement::remove_child(UIElement* child) 
{
	if (child == this)
	{
		return;
	}

	if (child->get_parent() != this)
	{
		return;
	}

	auto it = std::find(
		this->children.begin(),
		this->children.end(),
		child
	);

	if (it != this->children.end())
	{
		this->children.erase(it);

		child->parent = nullptr;

		child->on_removed_from_tree();

		this->internal_remove_child(child);
	}
}

UIElement* UIElement::get_parent() const 
{
	return this->parent;
}

const std::vector<UIElement*>& UIElement::get_children() const 
{
	return this->children;
}

void UIElement::set_parent(UIElement* parent) 
{
	if (parent == this)
	{
		return;
	}

	if (this->parent == parent)
	{
		return;
	}

	if (this->parent)
	{
		this->parent->remove_child(this);
	}

	if (parent)
	{
		parent->add_child(this);
	}
}

void UIElement::internal_draw() {}
void UIElement::internal_update() {}

void UIElement::draw() 
{
	for (UIElement* child : this->children)
	{
		child->draw();
	}

	this->internal_draw();
}

void UIElement::update()
{
	this->process_queue();

	if (this->parent)
	{
		this->global_offset = this->offset + this->parent->get_global_offset();
	}
	else
	{
		this->global_offset = this->offset;
	}

	for (UIElement* child : this->children)
	{
		child->update();
	}

	this->internal_update();
}

void UIElement::internal_add_child(UIElement* child) {}
void UIElement::internal_remove_child(UIElement* child) {}

void UIElement::on_destroyed() {}

sf::Vector2f UIElement::get_offset() const
{
	return this->offset;
}
void UIElement::set_offset(const sf::Vector2f& offset)
{
	this->offset = offset;
}

sf::Vector2f UIElement::get_global_offset() const
{
	return this->global_offset;
}

sf::RectangleShape& UIElement::get_size()
{
	return this->size;
}
void UIElement::deferred_add_child(UIElement* child) 
{
	auto it = std::find(
		this->add_queue.begin(),
		this->add_queue.end(),
		child
	);

	if (it != this->add_queue.end())
	{
		return;
	}

	this->add_queue.emplace_back(child);
}

void UIElement::deferred_remove_child(UIElement* child) 
{
	auto it = std::find(
		this->remove_queue.begin(),
		this->remove_queue.end(),
		child
	);

	if (it != this->remove_queue.end())
	{
		return;
	}

	this->remove_queue.emplace_back(child);
}

void UIElement::process_queue()
{
	for (UIElement* to_be_added : this->remove_queue)
	{
		auto it = std::find(
			this->children.begin(),
			this->children.end(),
			to_be_added
		);

		if (it != this->children.end())
		{
			this->children.erase(it);
		}
	}

	for (UIElement* to_be_added : this->add_queue)
	{
		auto it = std::find(
			this->children.begin(),
			this->children.end(),
			to_be_added
		);

		if (it == this->children.end())
		{
			this->children.emplace_back(to_be_added);
		}
	}

	this->remove_queue.clear();
	this->add_queue.clear();
}

UITree* UIElement::get_tree() const
{
	return this->tree;
}

File: Tev/UI/UIElement.cs
#pragma warning disable IDE0130
#pragma warning disable IDE0031

using System;
using System.Collections.Generic;
using Microsoft.Xna.Framework;
using MoonSharp.Interpreter;
using Tev.Core;

namespace Tev.UI;

public class UIElement
{
    [MoonSharpHidden]
    public Vector2 Offset { get; set; }

    public uint LayerDepth { get; set; } = 0;

    protected float internalLayerDepth = 0;

    private UIElement? parent;
    private readonly UITreeService tree;

    private readonly HashSet<UIElement> children;

    private readonly HashSet<Action> doQueue;

    [MoonSharpHidden]
    public Vector2 GlobalOffset { get; private set; }

    public const float LayerDepthDivider = 10000000;



    public UIElement(UITreeService tree)
    {
        children = [];
        doQueue = [];
        this.tree = tree;
    }

    public virtual void OnAddedToTree() { }
    public virtual void OnRemovedFromTree() { }


    protected virtual void InternalUpdate(GameTime gameTime)
    {

    }

    protected virtual void InternalDraw(GameTime gameTime)
    {

    }

    [MoonSharpHidden]
    public HashSet<Action> GetDoQueue() => doQueue;

    [MoonSharpHidden]
    public void Update(GameTime gameTime)
    {
        ProcessQueue();

        if (parent != null)
        {
            GlobalOffset = Offset + parent.GlobalOffset;
        }
        else
        {
            GlobalOffset = Offset;
        }

        internalLayerDepth = ((float)LayerDepth / LayerDepthDivider);

        foreach (var child in children)
        {
            child.Update(gameTime);
        }

        InternalUpdate(gameTime);
    }

    [MoonSharpHidden]
    public void Draw(GameTime gameTime)
    {
        foreach (var child in children)
        {
            child.Draw(gameTime);
        }

        InternalDraw(gameTime);
    }


    public void AddChild(UIElement child)
    {
        if (child.parent != null)
        {
            Console.WriteLine("Element already has parent!");
            return;
        }

        children.Add(child);

        child.parent = this;

        child.OnAddedToTree();
    }

    public void RemoveChild(UIElement child)
    {
        if (child.parent != this)
        {
            Console.WriteLine("Element's parent is NOT ME!");
            return;
        }

        children.Remove(child);

        child.parent = null;

        child.OnRemovedFromTree();
    }

    public UIElement? GetParent() => parent;

    public UITreeService GetTree() => tree;

    public void SetParent(UIElement other)
    {
        if (children.Contains(other) || this == other)
        {
            Console.WriteLine("Cannot set parent to child or self!");
            return;
        }

        if (parent != null)
        {
            parent.RemoveChild(this);
        }

        parent = other;

        if (parent != null)
        {
            parent.AddChild(this);
        }
    }

    private void ProcessQueue()
    {
        foreach (var action in doQueue)
        {
            action?.Invoke();
        }

        doQueue.Clear();
    }

    public void DeferredAddChild(UIElement child)
    {
        if (child.parent != null)
        {
            Console.WriteLine("Element already has parent!");
            return;
        }

        doQueue.Add(() => AddChild(child));
    }

    public void DoDeferred(Action queuedAction)
        => doQueue.Add(queuedAction);

    public void DeferredRemoveChild(UIElement child)
    {
        if (child.parent != this)
        {
            Console.WriteLine("Element's parent is NOT ME!");
            return;
        }

        doQueue.Add(() => RemoveChild(child));
    }

    [MoonSharpHidden]
    public HashSet<UIElement> GetChildren() => children;
}